### DISCUSSION TITLE
Servlet Filter with ThreadLocal in DGS

### DISCUSSION THREAD

**Discussion by vzhemevko** [Netflix/dgs-framework#570] [Q&A] [answered]
Stack:
Spring Boot: 2.4.5
DGS: 3.11.0 

Case: 
There is a need to filter each request and read from it HTTP headers. The value from the headers should be accessible in downstream code which processes the GraphQL request. A common approach could be to use the Servlet filter with ThreadLocal for storing those values. 

Issue: 
DGS is using CompletableFutures to process the incoming GraphQL requests, so even if the filter is set up and the data is read to the ThreadLocal it won't be available in the execution thread of DGS.

**What would be recommended ways to resolve this kind of issue?**

**Comment by paulbakker** [comment, marked as answer]
You use the custom context mechanism for this: https://netflix.github.io/dgs/datafetching/#using-context, specifically the `DgsCustomContextBuilderWithRequest`, which gives access to the headers in the builder.
